master/buildbot/process/buildrequestmerger.py

```diff
@@ -0,0 +1,246 @@
+import json
+import time
+from collections import defaultdict
+
+from twisted.application import service
+from twisted.internet import reactor
+from twisted.python import log
+import sqlalchemy as sa
+
+from buildbot import config
+from buildbot.util.lru import LRUCache
+
+
+class BuildRequestMerger(config.ReconfigurableServiceMixin, service.Service):
+    """
+    Wrapper around buildsets.addBuildset that does merging before
+    buildrequests hit the db.
+
+    This will only do merges within a same buildchain, and will never merge
+    the top level build.
+    """
+
+    # Basic list of properties that must match for a buildrequest to be merged
+    # BuilderConfigs can define additional properties (see _propertiesMatch)
+    BASE_MERGE_PROPERTIES = ['force_rebuild', 'force_chain_rebuild']
+
+    def __init__(self, master):
+        self.master = master
+        self.startbrid_cache = LRUCache(
```

Comment:
I think the async version of the cache would perform better for this case, AsyncLRUCache  

> class AsyncLRUCache(LRUCache):
>     """
>     An LRU cache with asynchronous locking to ensure that in the common case of
>     multiple concurrent requests for the same key, only one fetch is performed.
>     """

Reply:
The problem is that I'm updating the cache manually (by reading multiple properties at once), and doing that in the AsyncLRU seemed too finicky/unsafe.

We could:
* Rewrite some of the AsyncLRU to allow us to add multiple items in one move
* Revert back to reading one cache entry from the DB at the time (using the standard miss function flow)
* Take the performance hit and use regular LRU (what I'm doing right now)

What do you think?

Reply:
@campos-ddc I think we can make `self.startbrid_cache` use AsyncLRUCache  and I think if we make sure that the `self.properties_cache` gets called in an async way, it should be ok. Probably we would need `_getMergeBrid` used with `@defer.inlineCallbacks`